File: src/doctor.ts
import { readdir, readFile } from "node:fs/promises";
import { join } from "node:path";
import { isNoteShaped, serializeNote } from "./notes.ts";
import type { DataPaths } from "./paths.ts";

export interface DoctorReport {
  issues: string[];
  checkedFiles: number;
}

async function readOrNull(path: string, label: string, issues: string[]): Promise<string | null> {
  try {
    return await readFile(path, "utf-8");
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code !== "ENOENT") {
      issues.push(`${label}: unreadable (${code ?? (err as Error).message})`);
    }
    return null;
  }
}

async function listDir(dir: string, label: string, issues: string[]): Promise<string[]> {
  try {
    return await readdir(dir);
  } catch (err: unknown) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code !== "ENOENT") {
      issues.push(`${label}: unreadable (${code ?? (err as Error).message})`);
    }
    return [];
  }
}

const STROKE_FIELDS = ["t", "d", "p", "spm", "hr"] as const;

function isStrokeShaped(parsed: unknown): boolean {
  if (parsed == null || typeof parsed !== "object" || Array.isArray(parsed)) return false;
  const row = parsed as Record<string, unknown>;
  return STROKE_FIELDS.every((k) => row[k] === undefined || typeof row[k] === "number");
}

export async function runDoctor(paths: DataPaths): Promise<DoctorReport> {
  const issues: string[] = [];
  let checkedFiles = 0;

  const meta = await readOrNull(paths.meta, "meta.json", issues);
  if (meta != null) {
    checkedFiles++;
    try {
      const parsed = JSON.parse(meta);
      if (typeof parsed?.schema_version !== "number") {
        issues.push("meta.json: missing numeric schema_version");
      }
    } catch {
      issues.push("meta.json: not valid JSON");
    }
  }

  const workouts = await readOrNull(paths.workouts, "workouts.jsonl", issues);
  if (workouts != null) {
    checkedFiles++;
    let lineNo = 0;
    for (const line of workouts.split("\n")) {
      lineNo++;
      if (line.trim() === "") continue;
      let parsed: unknown;
      try {
        parsed = JSON.parse(line);
      } catch {
        issues.push(`workouts.jsonl: line ${lineNo} is not valid JSON`);
        continue;
      }
      const w = parsed as { id?: unknown; date?: unknown; distance?: unknown; time?: unknown };
      if (
        parsed == null ||
        typeof parsed !== "object" ||
        typeof w.id !== "number" ||
        typeof w.date !== "string" ||
        typeof w.distance !== "number" ||
        typeof w.time !== "number"
      ) {
        issues.push(`workouts.jsonl: line ${lineNo} malformed workout record`);
      }
    }
  }

  for (const f of await listDir(paths.strokesDir, "strokes/", issues)) {
    if (!f.endsWith(".jsonl")) continue;
    const text = await readOrNull(join(paths.strokesDir, f), `strokes/${f}`, issues);
    if (text == null) continue;
    checkedFiles++;
    let lineNo = 0;
    for (const line of text.split("\n")) {
      lineNo++;
      if (line.trim() === "") continue;
      let parsed: unknown;
      try {
        parsed = JSON.parse(line);
      } catch {
        issues.push(`strokes/${f}: line ${lineNo} is not valid JSON`);
        continue;
      }
      if (!isStrokeShaped(parsed)) {
        issues.push(`strokes/${f}: line ${lineNo} malformed stroke record`);
      }
    }
  }

  const looseContent = new Map<string, { content: string; file: string }>();
  const divergentIds = new Set<string>();
  for (const f of await listDir(paths.notesDir, "notes/", issues)) {
    if (!f.endsWith(".json")) continue;
    const text = await readOrNull(join(paths.notesDir, f), `notes/${f}`, issues);
    if (text == null) continue;
    checkedFiles++;
    try {
      const parsed = JSON.parse(text) as unknown;
      if (!isNoteShaped(parsed)) {
        issues.push(`notes/${f}: malformed note record`);
      } else {
        const content = serializeNote(parsed);
        const prior = looseContent.get(parsed.id);
        if (prior != null && prior.content !== content && !divergentIds.has(parsed.id)) {
          divergentIds.add(parsed.id);
          issues.push(
            `notes: divergent copies of note ${parsed.id} (${prior.file}, ${f}); reconcile before they compact`,
          );
        }
        looseContent.set(parsed.id, { content, file: f });
      }
    } catch {
      issues.push(`notes/${f}: not valid JSON`);
    }
  }

  const archiveIds = new Set<string>();
  for (const f of await listDir(paths.archiveDir, "notes/archive/", issues)) {
    if (!f.endsWith(".jsonl")) continue;
    const text = await readOrNull(join(paths.archiveDir, f), `notes/archive/${f}`, issues);
    if (text == null) continue;
    checkedFiles++;
    let prevMs = Number.NEGATIVE_INFINITY;
    let prevId = "";
    let lineNo = 0;
    for (const line of text.split("\n")) {
      lineNo++;
      if (line.trim() === "") continue;
      let parsed: unknown;
      try {
        parsed = JSON.parse(line);
      } catch {
        issues.push(`notes/archive/${f}: line ${lineNo} is not valid JSON`);
        continue;
      }
      if (!isNoteShaped(parsed)) {
        issues.push(`notes/archive/${f}: line ${lineNo} malformed note record`);
        continue;
      }
      const ms = new Date(parsed.date).getTime();
      if (ms < prevMs || (ms === prevMs && parsed.id < prevId)) {
        issues.push(`notes/archive/${f}: line ${lineNo} out of (date, id) order`);
      }
      prevMs = ms;
      prevId = parsed.id;
      if (archiveIds.has(parsed.id)) {
        issues.push(`notes/archive/${f}: duplicate note id ${parsed.id}`);
      }
      archiveIds.add(parsed.id);
    }
  }

  return { issues, checkedFiles };
}

File: internal/doctor/doctor.go
package doctor

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/richhaase/c2/internal/notes"
	"github.com/richhaase/c2/internal/paths"
	"github.com/richhaase/c2/internal/storage"
)

type Report struct {
	Issues       []string
	CheckedFiles int
}

var strokeFields = []string{"t", "d", "p", "spm", "hr"}

type checker struct {
	report *Report
}

func (c *checker) issue(format string, args ...any) {
	c.report.Issues = append(c.report.Issues, fmt.Sprintf(format, args...))
}

func (c *checker) readOrNil(path, label string) []byte {
	data, err := os.ReadFile(path)
	if err != nil {
		if !storage.IsMissing(err) {
			c.issue("%s: unreadable (%v)", label, err)
		}
		return nil
	}
	return data
}

func (c *checker) listDir(dir, label string) []os.DirEntry {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if !storage.IsMissing(err) {
			c.issue("%s: unreadable (%v)", label, err)
		}
		return nil
	}
	return entries
}

func isStrokeShaped(raw []byte) bool {
	var parsed map[string]any
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return false
	}
	for _, key := range strokeFields {
		value, present := parsed[key]
		if !present || value == nil {
			continue
		}
		if _, ok := value.(float64); !ok {
			return false
		}
	}
	return true
}

func nonEmptyLines(data []byte) [][]byte {
	var out [][]byte
	for line := range bytes.SplitSeq(data, []byte("\n")) {
		out = append(out, line)
	}
	return out
}

func Run(p paths.DataPaths) Report {
	report := Report{Issues: []string{}}
	c := &checker{report: &report}

	c.checkMeta(p)
	c.checkWorkouts(p)
	c.checkStrokes(p)
	c.checkLooseNotes(p)
	c.checkArchives(p)

	return report
}

func (c *checker) checkMeta(p paths.DataPaths) {
	data := c.readOrNil(p.Meta, "meta.json")
	if data == nil {
		return
	}
	c.report.CheckedFiles++
	var parsed map[string]any
	if err := json.Unmarshal(data, &parsed); err != nil {
		c.issue("meta.json: not valid JSON")
		return
	}
	if _, ok := parsed["schema_version"].(float64); !ok {
		c.issue("meta.json: missing numeric schema_version")
	}
}

func (c *checker) checkWorkouts(p paths.DataPaths) {
	data := c.readOrNil(p.Workouts, "workouts.jsonl")
	if data == nil {
		return
	}
	c.report.CheckedFiles++
	for i, line := range nonEmptyLines(data) {
		lineNo := i + 1
		if len(bytes.TrimSpace(line)) == 0 {
			continue
		}
		var parsed map[string]any
		if err := json.Unmarshal(line, &parsed); err != nil {
			c.issue("workouts.jsonl: line %d is not valid JSON", lineNo)
			continue
		}
		_, idOK := parsed["id"].(float64)
		_, dateOK := parsed["date"].(string)
		_, distanceOK := parsed["distance"].(float64)
		_, timeOK := parsed["time"].(float64)
		if !idOK || !dateOK || !distanceOK || !timeOK {
			c.issue("workouts.jsonl: line %d malformed workout record", lineNo)
		}
	}
}

func (c *checker) checkStrokes(p paths.DataPaths) {
	for _, entry := range c.listDir(p.StrokesDir, "strokes/") {
		name := entry.Name()
		if !strings.HasSuffix(name, ".jsonl") {
			continue
		}
		label := "strokes/" + name
		data := c.readOrNil(filepath.Join(p.StrokesDir, name), label)
		if data == nil {
			continue
		}
		c.report.CheckedFiles++
		for i, line := range nonEmptyLines(data) {
			lineNo := i + 1
			if len(bytes.TrimSpace(line)) == 0 {
				continue
			}
			if !json.Valid(line) {
				c.issue("%s: line %d is not valid JSON", label, lineNo)
				continue
			}
			if !isStrokeShaped(line) {
				c.issue("%s: line %d malformed stroke record", label, lineNo)
			}
		}
	}
}

func (c *checker) checkLooseNotes(p paths.DataPaths) {
	type seenNote struct {
		content string
		file    string
	}
	looseContent := map[string]seenNote{}
	divergent := map[string]bool{}

	for _, entry := range c.listDir(p.NotesDir, "notes/") {
		name := entry.Name()
		if !strings.HasSuffix(name, ".json") {
			continue
		}
		label := "notes/" + name
		data := c.readOrNil(filepath.Join(p.NotesDir, name), label)
		if data == nil {
			continue
		}
		c.report.CheckedFiles++
		if !json.Valid(data) {
			c.issue("%s: not valid JSON", label)
			continue
		}
		note, ok := notes.Parse(data)
		if !ok {
			c.issue("%s: malformed note record", label)
			continue
		}
		content := notes.Serialize(note)
		if prior, exists := looseContent[note.ID]; exists && prior.content != content && !divergent[note.ID] {
			divergent[note.ID] = true
			c.issue("notes: divergent copies of note %s (%s, %s); reconcile before they compact", note.ID, prior.file, name)
		}
		looseContent[note.ID] = seenNote{content: content, file: name}
	}
}

func (c *checker) checkArchives(p paths.DataPaths) {
	archiveIDs := map[string]bool{}

	for _, entry := range c.listDir(p.ArchiveDir, "notes/archive/") {
		name := entry.Name()
		if !strings.HasSuffix(name, ".jsonl") {
			continue
		}
		label := "notes/archive/" + name
		data := c.readOrNil(filepath.Join(p.ArchiveDir, name), label)
		if data == nil {
			continue
		}
		c.report.CheckedFiles++

		hasPrev := false
		var prevMs int64
		prevID := ""
		for i, line := range nonEmptyLines(data) {
			lineNo := i + 1
			if len(bytes.TrimSpace(line)) == 0 {
				continue
			}
			if !json.Valid(line) {
				c.issue("%s: line %d is not valid JSON", label, lineNo)
				continue
			}
			note, ok := notes.Parse(line)
			if !ok {
				c.issue("%s: line %d malformed note record", label, lineNo)
				continue
			}
			t, _ := notes.ParseDate(note.Date)
			ms := t.UnixMilli()
			if hasPrev && (ms < prevMs || (ms == prevMs && note.ID < prevID)) {
				c.issue("%s: line %d out of (date, id) order", label, lineNo)
			}
			hasPrev = true
			prevMs = ms
			prevID = note.ID
			if archiveIDs[note.ID] {
				c.issue("%s: duplicate note id %s", label, note.ID)
			}
			archiveIDs[note.ID] = true
		}
	}
}
